tests/ui/single_field_patterns.rs

```diff
@@ -0,0 +1,89 @@
+#![warn(clippy::single_field_patterns)]
+struct Struct {
+    field1: Option<i32>,
+    field2: Option<i32>,
+}
+
+fn lint_struct(struct1: Struct) {
+    let Struct { field1, .. } = struct1;
+    let Struct { field1, field2: _ } = struct1;
+    match struct1 {
+        Struct { field1: Some(n), .. } if n >= 50 => {},
+        Struct { field1: None, .. } => {},
+        _ => {},
+    }
+    match struct1 {
+        Struct { field1: Some(n), .. } if n >= 50 => {},
+        Struct { .. } => {},
+        _ => {},
+    }
+    match struct1 {
+        Struct { field1: Some(1), .. } => {},
+        Struct { field1: Some(2), .. } => {},
+        Struct { field1: None, .. } => {},
+        _ => {},
+    }
+    match struct1 {
+        Struct {
+            field1: Some(_) | None, ..
+        } => {},
+    }
+    while let Struct { field1: Some(5), .. } = struct1 {}
+}
+
+fn lint_ref(struct1: &&Struct) {
+    // this should suggest struct1.field1, NOT **struct1.field1
+    let Struct { field1, .. } = **struct1;
```

Comment:
Please also add a test case where the field is `ref`ed:

```suggestion
    let Struct { ref field1, .. } = **struct1;
```

or `ref mut`ed